README.txt
...One-way coupled FNPT-RANS model for wave-structure interaction problems...
FNPT ----> IITM-FNPT2D
RANS ----> IITM-RANS3D

Code is currently setup for simulating the interaction between a H/d=0.6 solitary wave (d=1m) and a circular cylinder of diameter 1m in a 20m x 4m x 2m. 
The domain is discretized using a uniform mesh of 500 x 50 x 100 (NX x NY x NZ). 

ALL CODES ARE CURRENTLY SERIAL. The entire simulation involves sequential execution of FIVE different codes.  

Linux is recommended and necessary for fast execution of all FIVE codes. In addition, the executable "fnpt" located in "IITM-FNPT2D" will only run in Linux.  

Linux command for compiling any code in terminal: g++ <CODENAME> -O2
Linux command for running the compiled code in terminal: ./a.out
	Note: Prof. Sriram's FNPT code is already compiled as an executable in the folder "IITM-FNPT2D" so compilation is not required in this case. Execute in terminal using the command: ./fnpt [will only run in Linux] 

Steps to be followed for the SOLITON-CYLINDER INTERACTION simulation:

(1) Go to the folder: "PADDLE-SIGNAL" ; compile and run the code "SOLITARY_WAVE_PADDLE_DISP.CPP". After execution, the code will output "wavePaddle.dat" which is the piston-type wave-paddle displacement history required for soliton generation. Open "wavePaddle.dat" in a text editor, enable display of linenumbers, note the total number of lines in the file and write that number at the top of the file in a newline and save. Alternately "wavePaddle.dat" can also be created using Prof. Sriram's MATLAB code "WaveGen". 
		Note: In "SOLITARY_WAVE_PADDLE_DISP.CPP", the file "wavePaddle.dat" gets written in "a" or "append" mode. So if you make changes in "SOLITARY_WAVE_PADDLE_DISP.CPP" pls. delete the previous "wavePaddle.dat" before re-running the code.  

(2) Copy "wavePaddle.dat" to the folder: "IITM-FNPT2D" and run the compiled executable "fnpt". The simulation is setup by changing the parameters in "FEMINPUT.dat". It's currently setup for H/d=0.6 soliton generation. Once execution of "fnpt" is complete, focus on two files: 
	(a) "Output_bvout.dat" - check surface-elevation time series at different wave-probe locations as specified in "FEMINPUT.dat" ; 
	(b) "Output_PHIT.dat" - contains the Eulerian wave-field (pressure, velocity, acceleration etc.) which is to be input to "IITM-RANS3D" after suitable interpolation. 
		Note: Input time-series in terms of paddle displacement is mandatory for ANY type of wave to be generated using "IITM-FNPT2D". Always naming the time-series file "wavePaddle.dat" is advised. 
Critical: [only applicable to this H/d=0.6 soliton] - Given the strong non-linearity of the soliton, the "fnpt" run (see step (2) above) will PAUSE at: "RUNNING TIMESTEP = 5513". Do not force continue the run beyond this point. Press "CTRL+C" in the terminal to exit. 

(3) Copy "Output_PHIT.dat" to the folder: "NSE-INPUT-EXTRACTION". Compile and run the code "FNPT_to_NS.CPP". After execution is complete, focus on three files: 
	(a) "Interp_pf.dat" - interpolated spatio-temporal series of the pressure and volume-fraction field [COLLOCATED MESH] ; 
	(b) "Interp_Ustag.dat" - interpolated spatio-temporal series of the U-velocity field [STAGGERED] ;
	(c) "Interp_Wstag.dat" - interpolated spatio-temporal series of the W-velocity field [STAGGERED].

(4) Copy "Interp_pf.dat", "Interp_Ustag.dat" and "Interp_Wstag.dat" to the folder: "IITM-RANS3D". 

(5) Go to the folder: "SOLID-PATCH" ; compile and run the code "VOF_INIT.c". Once execution is complete, the code will output two files: 
	(a) "CYLINDER.dat" - contains the entire SOLID_VOF field for patching the cylinder within the RANS domain
	(b) "VOFINIT_tecplot.dat" - contains the same information as "CYLINDER.dat" but this file can be imported in TecPlot/ParaVIEW to check quality/correctness of the initial patch. 
		Note: Inclusion of the -O2 optimization flag during compilation is essential to ensure rapid execution of "VOF_INIT.c". 

(6) Copy "CYLINDER.dat" to the folder: "IITM-RANS3D". Compile and run "3-D_TWO-PHASE-NSE.CPP". Compilation using the -O2 flag is advisable for faster execution. Invoking the -O2 flag during compilation will flash some WARNINGS related to reading from the files "Interp_pf.dat", "Interp_Ustag.dat", "Interp_Wstag.dat" and "CYLINDER.dat" - just ignore these warnings. Import the output files: "000_1.dat", "000_2.dat"....etc. to TecPlot/ParaVIEW for analyzing the 3D-RANS results. 
	Note: (a) "3-D_TWO-PHASE-NSE.CPP" is the main code - all functions and variable declarations are available in "functions.h".
		  (b) The global RANS mesh "NX x NY x NZ" is defined in "functions.h" via the pre-processor directive: #define ; for intance: "#define NY 50". Change the global cell-count in "functions.h" as per requirement. 
		  (c) Everything else in "functions.h" is hard-coded - changing any function/variable-declaration other than the global cell-count (NX x NY x NZ) may result in unpredictable behavior/incorrect solution.
Critical: ENSURE CONSISTENCY IN NS MESH-SIZE ACROSS STEPS (2)-(6). Directly changing NX or NZ in step (6) will lead to a wrong result.  

Contact Dr. Shaswat Saincher for any queries/issues at: shaswat.saincher@gmail.com  

CHANGELOG.txt
03rd March 2021......................................................................................................................................
Code in question:				3-D_TWO-PHASE-NSE.CPP
Description of error:			VOF boundary condition error at lines "513" and "777"
Existing line:					f[i][IMG-1][k]=2.0*f[i][IMG][k]-w[i][IMG+1][k];
Corrected to:					f[i][IMG-1][k]=2.0*f[i][IMG][k]-f[i][IMG+1][k];

26th May 2021........................................................................................................................................
Code in question: 				3-D_TWO-PHASE-NSE.CPP
Description of error:			Whilst global Courant number is restricted to never exceed C_PERMIT(=0.2 generally)...the RANS time-step-size (dt) may EXCEED the FNPT time-step-size (dt_FNPT) which would cause an error in interpolation. So the variable time-stepping routine (lines "259" to "269") has been modified as follows:
Existing lines:
//VARIABLE TIME-STEPPING (dt=init_time_step unless Courant exceeds C_PERMIT)
if (trigger==0 && maxcour<C_PERMIT)								//keep equal to the initial time-step until the trigger is activated
	dt=init_time_step;
else
{
	if (maxcour<C_PERMIT)
	dt=1.01*dt;													//don't directly revert back to "init_time_step" as "maxcour" may exceed "C_PERMIT"...rather ramp up "dt" slowly
	else
	dt=C_PERMIT*mincel/maxvel;
	trigger=1;
}
Corrected to:
//VARIABLE TIME-STEPPING (dt=init_time_step unless Courant exceeds C_PERMIT)
if (trigger==0 && maxcour<C_PERMIT)								//keep equal to the initial time-step until the trigger is activated
	dt=init_time_step;
else
{
	if (maxcour<C_PERMIT)
	{
	dt=1.01*dt;													//don't directly revert back to "init_time_step" as "maxcour" may exceed "C_PERMIT"...rather ramp up "dt" slowly
		if (dt>init_time_step)
		dt=init_time_step;										//ensure that NS time-step size does not exceed dt_FNPT
	}
	else
	{
	dt=C_PERMIT*mincel/maxvel;
		if (dt>init_time_step)
		dt=init_time_step;										//ensure that NS time-step size does not exceed dt_FNPT
	}
	trigger=1;
}




4.SOLID-PATCH/VOF_INIT.c
#include <stdio.h>
#include <math.h>
#include <time.h>

#define NX 500											/****************************************************/
#define NY 50											/*...has to be consistent with the Global NS mesh...*/
#define NZ 100											/****************************************************/

int i,j,k;
double L,B,H,dx,dy,dz,px[NX+10],py[NY+10],pz[NZ+10],cx[NX+10],cy[NY+10],cz[NZ+10];
double lx[NX+10],ly[NY+10],lz[NZ+10],rx,ry,rz,exsum;
int IMG=4;											//number of imaginary cells - has to be consistent with "3-D_TWO-PHASE-NSE.CPP"
double f[NX+10][NY+10][NZ+10]={0.0};

FILE *f1;

double pi=3.141592653589793238462643383279502884197;
double a_cyl=10.0,b_cyl=2.0,r=0.5;								//cylinder centered at (x=a_cyl,y=b_cyl) having a radius of "r"

main()
{
clock_t tt;
tt=clock();

printf("This is a code to PATCH the FFD VOF-field for a circular cylinder\n\n");

L=20.0;												//length of the NS domain
B=4.0;												//width of the NS domain
H=2.0;												//height of the NS domain

rx=1.0;												//mesh-stretching parameter along x
ry=1.0;												//mesh-stretching parameter along y
rz=1.0;												//mesh-stretching parameter along z

//computing cell-sizes along the x-direction
exsum=1.0;
for (i=1;i<=(NX-1);i++)
    exsum=exsum+pow(rx,i);

lx[IMG+1]=L/exsum;										//first length in x
for (i=IMG+2;i<=NX+IMG;i++)
    lx[i]=lx[i-1]*rx;

//computing cell-sizes along the y-direction
exsum=1.0;
for (j=1;j<=(NY-1);j++)
    exsum=exsum+pow(ry,j);

ly[IMG+1]=B/exsum;										//first length in y
for (j=IMG+2;j<=NY+IMG;j++)
    ly[j]=ly[j-1]*ry;

//computing cell-sizes along the z-direction
exsum=1.0;
for (k=1;k<=(NZ-1);k++)
    exsum=exsum+pow(rz,k);

lz[IMG+1]=H/exsum;										//first length in z
for (k=IMG+2;k<=NZ+IMG;k++)
    lz[k]=lz[k-1]*rz;

//calculating the coordinates of cell faces
px[IMG+1]=0.0;
py[IMG+1]=0.0;
pz[IMG+1]=0.0;
for (i=IMG+2;i<=NX+IMG+1;i++)
    px[i]=px[i-1]+lx[i-1];
for (j=IMG+2;j<=NY+IMG+1;j++)
    py[j]=py[j-1]+ly[j-1];
for (k=IMG+2;k<=NZ+IMG+1;k++)
    pz[k]=pz[k-1]+lz[k-1];

//calculating the coordinates of cell centers
cx[IMG+1]=0.5*lx[IMG+1];
cy[IMG+1]=0.5*ly[IMG+1];
cz[IMG+1]=0.5*lz[IMG+1];
for (i=IMG+2;i<=NX+IMG+1;i++)
    cx[i]=cx[i-1]+0.5*(lx[i-1]+lx[i]);
for (j=IMG+2;j<=NY+IMG+1;j++)
    cy[j]=cy[j-1]+0.5*(ly[j-1]+ly[j]);
for (k=IMG+2;k<=NZ+IMG+1;k++)
    cz[k]=cz[k-1]+0.5*(lz[k-1]+lz[k]);

//pixelization routine for patching
for (i=IMG+1;i<=NX+IMG;i++)
{
    for (j=IMG+1;j<=NY+IMG;j++)
    {
        for (k=IMG+1;k<=NZ+IMG;k++)								//cylinder spans the entire height of the domain
        {
            /*...Checking whether ANY of the 8 hexahedron vertices fall INSIDE the cylinder...*/
	    if ((pow((px[i]-a_cyl),2.0)+pow((py[j]-b_cyl),2.0)<r*r) || (pow((px[i+1]-a_cyl),2.0)+pow((py[j]-b_cyl),2.0)<r*r) || (pow((px[i]-a_cyl),2.0)+pow((py[j+1]-b_cyl),2.0)<r*r) || (pow((px[i+1]-a_cyl),2.0)+pow((py[j+1]-b_cyl),2.0)<r*r))
                {
                    /*...Checking whether ALL of the 8 hexahedron vertices fall INSIDE the cylinder...*/
		    if ((pow((px[i]-a_cyl),2.0)+pow((py[j]-b_cyl),2.0)<r*r) && (pow((px[i+1]-a_cyl),2.0)+pow((py[j]-b_cyl),2.0)<r*r) && (pow((px[i]-a_cyl),2.0)+pow((py[j+1]-b_cyl),2.0)<r*r) && (pow((px[i+1]-a_cyl),2.0)+pow((py[j+1]-b_cyl),2.0)<r*r))
                        f[i][j][k]=1.0;								//pure cell (VOF_SOLID=1)
                    else									//mixed cell (0<VOF_SOLID<1)
                    {
                        int l,m,n,VOFcount=0,SUBcells=200;
                        double dxx,dyy,dzz;
                        dxx=lx[i]/SUBcells;
                        dyy=ly[j]/SUBcells;
                        dzz=lz[k]/SUBcells;
			//generating a SUBcells^3 sub-cell mesh (face-centered)
			double pxx[SUBcells+10],pyy[SUBcells+10],pzz[SUBcells+10];
			    pxx[1]=px[i];
			    pyy[1]=py[j];
			    pzz[1]=pz[k];
			for (l=2;l<=SUBcells+1;l++)
			    pxx[l]=pxx[l-1]+dxx;
			for (m=2;m<=SUBcells+1;m++)
			    pyy[m]=pyy[m-1]+dyy;
			for (n=2;n<=SUBcells+1;n++)
			    pzz[n]=pzz[n-1]+dzz;

                            for (l=1;l<=SUBcells;l++)
                            {
                                for (m=1;m<=SUBcells;m++)
                                {
                                    for (n=1;n<=SUBcells;n++)
                                    {
				    /*...Checking whether ALL of the 8 SUBcell hexahedron vertices fall INSIDE the cylinder...*/
				    if ((pow((pxx[l]-a_cyl),2.0)+pow((pyy[m]-b_cyl),2.0)<r*r) && (pow((pxx[l+1]-a_cyl),2.0)+pow((pyy[m]-b_cyl),2.0)<r*r) && (pow((pxx[l]-a_cyl),2.0)+pow((pyy[m+1]-b_cyl),2.0)<r*r) && (pow((pxx[l+1]-a_cyl),2.0)+pow((pyy[m+1]-b_cyl),2.0)<r*r))
                                        VOFcount=VOFcount+1;
                                    }
                                }
                            }
                        f[i][j][k]=(double)VOFcount/(double)(pow(SUBcells,3.0));
                        printf("The code is here: i=%d\tj=%d\tk=%d\tf=%f\tVOFcount=%d\n",i,j,k,f[i][j][k],VOFcount);
                    }
                }
            else										//pure cell (VOF_SOLID=0)
                continue;
        }
    }
}

//writing the initialized/patched VOF field to file - TO BE READ BY IITM-RANS3D
f1=fopen("CYLINDER.dat","w");
for (i=IMG+1;i<=NX+IMG;i++)
    for (j=IMG+1;j<=NY+IMG;j++)
        for (k=IMG+1;k<=NZ+IMG;k++)
        fprintf(f1,"%f\t%f\t%f\t%0.10f\n",cx[i],cy[j],cz[k],f[i][j][k]);
fclose(f1);

//writing the initialized/patched VOF field to file - FOR CHECKING QUALITY OF THE PATCH IN TECPLOT/PARAVIEW
f1=fopen("VOFINIT_tecplot.dat","w");
fprintf (f1,"\nVARIABLES = ""X"", ""Y"", ""Z"", ""VOF""\n");
fprintf (f1,"ZONE I=%d, J=%d, K=%d, ZONETYPE=ORDERED, DATAPACKING=POINT\n",NZ,NY,NX);
for (i=IMG+1;i<=NX+IMG;i++)
    for (j=IMG+1;j<=NY+IMG;j++)
        for (k=IMG+1;k<=NZ+IMG;k++)
        fprintf(f1,"%f\t%f\t%f\t%0.10f\n",cx[i],cy[j],cz[k],f[i][j][k]);
fclose(f1);

tt=clock()-tt;
printf("The computation time is %f seconds\n",(float)tt/CLOCKS_PER_SEC);
}


1.PADDLE-SIGNAL/SOLITARY_WAVE_PADDLE_DISP.CPP
/**************************************************************************************************************************/
/*THIS CODE CALCULATES THE TOPOLOGY AS WELL AS PADDLE DISPLACEMENT CHARACTERISTICS REQUIRED FOR GENERATING A SOLITARY WAVE*/
/***********...GORING'S THEORY IS USED WHERE THE SOLITON HAS A HEIGHT "H" PROPAGATING IN A WATER DEPTH OF "d"...***********/
/**************************************************************************************************************************/

#include <stdio.h>
#include <math.h>

main()
{
double H=0.6;                               			//wave-height (m)
double d=1.0;                               			//water-depth (m)
double g=9.81;                              			//acceleration due to gravity
double C=sqrt(g*(H+d));                     			//phase velocity (m/s)
double k=sqrt((3.0*H)/(4.0*d*d*d));         			//wavenumber "k" (m^(-1))
double S=2.0*H/(k*d);                       			//required wave-maker stroke (m)
double T=(2.0*(3.8+(H/d)))/(k*C);           			//duration of soliton generation (s)
double T_init=2.0;                          			//initial period of inactivity (s)
double T_final=20.0;                       		 	//final period of inactivity (s)
double dt=0.01;                             			//step-size or sampling-rate

FILE *f1;                                   			//declaration of file-pointer
f1=fopen("wavePaddle.dat","a");             			//opening in a datafile in "append" mode

/*generation of time axis for the initial period of inactivity*/
double t_init[50000]={0.0};                 			//declaration of time array
int N_init=T_init/dt;                      			//number of time-steps required
t_init[0]=0.0;                              			//initial time
for (int i=1; i<=N_init; i++)
    t_init[i]=t_init[i-1]+dt;
for (int i=0; i<=N_init-1; i++)
    fprintf(f1,"%f\t%f\n",t_init[i],0.0);

/*generation of the time axis for period of soliton generation*/
double t[50000]={0.0};                      			//declaration of the time array
int N=T/dt;                                 			//number of time-steps required
t[0]=0.0;                                   			//initial time
for (int i=1; i<=N; i++)
    t[i]=t[i-1]+dt;                         			//generation of the time-vector

/*iterative calculation of the displacement*/
/*displacement formula modified so that WM only moves forwards in time*/
double X=0.0;                               			//initialization of wave-paddle displacement
double Xold=0.0;                            			//guessed value of paddle displacement
for (int i=0; i<=N; i++)                    			//loop over the entire time-vector
{
    do                                    			//simple RHS-LHS based iterative method
    {
        Xold=X;
        X=(H/(k*d))*(1.0+tanh(k*((C*(t[i]-0.5*T))-(Xold-0.5*S))));
        printf("Xold=%f\tX=%f\n",Xold,X);
    }while(fabs(X-Xold)>0.000000000001);

    do                                      			//Newton-Raphson method
    {
        Xold=X;
        X=Xold-(Xold-((H/(k*d))*(1.0+tanh(k*((C*(t[i]-0.5*T))-(Xold-0.5*S))))))/(1.0+(H/d)/(pow((cosh((k*C*(t[i]-0.5*T))-(k*(Xold-0.5*S)))),2.0)));
        printf("Xold=%f\tX=%f\n",Xold,X);
    }while(fabs(X-Xold)>0.000000000001);

    fprintf(f1,"%f\t%f\n",t[i]+T_init,X);
}

/*generation of time axis for the final period of inactivity*/
double t_final[50000]={0.0};                			//declaration of time array
int N_final=T_final/dt;
t_final[0]=T_init+t[N]+dt;
for (int i=1; i<=N_final; i++)
    t_final[i]=t_final[i-1]+dt;
for (int i=0; i<=N_final-1; i++)
    fprintf(f1,"%f\t%f\n",t_final[i],X);

fclose(f1);
}


Additional_SUBROUTINES/Mesh_Grading_calculator.CPP
//THIS CODE SOLVES FOR THE CELL-COUNT (nx) AND STRETCHING-FACTOR (rx) REQUIRED TO GENERATE A NON-UNIFORM MESH OVER A DOMAIN OF LENGTH "L"
//THE CODE IS ONLY INTENDED FOR A TYPICAL MESH GENERATION SCENARIO ENCOUNTERED WHERE ONLY THE DOMAIN-LENGTH, STARTING AND ENDING CELL-SIZES ARE KNOWN  
//FOLLOWING GEOMETRIC SERIES IS CONSIDERED: (1 + rx + rx^2 + rx^3 + .... + rx^(nx-1))*deltax_start = L ; where deltax_end = deltax_start*(rx^(nx-1)) 
//USE THE "-O2" OPTIMIZATION FLAG FOR FASTER EXECUTION

#include <stdio.h>
#include <math.h>

main()
{
//knowns
double L=4.0;											//length of the domain
double deltax_start=0.01;									//starting cell-size
double deltax_end=0.04;										//ending cell-size

//unknowns
double rx;											//mesh-stretching factor [UNKNOWN]
int nx;												//number of cells [UNKNOWN]

//bounds of cell-count
int nx_lower,nx_upper;										
if (deltax_start>deltax_end)									//if mesh is to be successively refined
	{
	nx_lower=L/deltax_start;								//uniformly coarse mesh throughout
	nx_upper=L/deltax_end;									//uniformly fine mesh throughout
	}
else												//if mesh is to be successively coarsened
	{
	nx_lower=L/deltax_end;									//uniformly coarse mesh throughout
	nx_upper=L/deltax_start;								//uniformly fine mesh throughout
	}
 
double tolerance=1e-03;										//error tolerance desired in the final solution
double mag_drx=1e-05;										//magnitude by which "rx" is incremented/decremented

//iterative solution routine for rx and nx
for (int count=nx_lower+1 ; count<=nx_upper-1 ; count++)					//loop between the lower and upper bounds for "nx"; excluding both values
{
printf("\nThe code is here..............count=%d............................\n",count);
rx=1.0;												// "rx" begins from unity and is incremented/decremented depending on the problem at hand
double drx=((deltax_end-deltax_start)/fabs(deltax_end-deltax_start))*(mag_drx);			//decides whether "rx" is to be incremented or decremented
double diff1=1.0,diff2=1.0;
double min1=1.0,min2=1.0,rxmin1,rxmin2;
double exsum;
	do
	{
	exsum=1.0;										//initiation of geometric progression
		diff1=deltax_start*pow(rx,count-1)-deltax_end;					//CONSTRAINT 1: relationship between first and last cell sizes
		{										//store the minimum value of "diff1"
			if (fabs(min1)>fabs(diff1))
			{
				min1=diff1;
				rxmin1=rx;		
			}
		}
		for (int i=1; i<=count-1; i++)
			exsum=exsum+pow(rx,i);							//building the geometric series
		diff2=L-exsum*deltax_start;							//CONSTRAINT 2: all cell-sizes SHOULD sum up to domain length	
		{										//store the minimum value of "diff2"
			if (fabs(min2)>fabs(diff2))
			{							
				min2=diff2;
				rxmin2=rx;
			}
		}
	if (rx<0.5 || rx>2.0)									//condition to prevent excessively stretched/compressed solutions
		break;
	else
		rx=rx+drx;
	}while((fabs(diff1)>tolerance) || (fabs(diff2)>tolerance)); 				//both constraints need to be simultaneously fulfilled	

	printf("count=%d \t diff1=%f \t rx1=%f \t diff2=%f \t rx2=%f\n",count,min1,rxmin1,min2,rxmin2);

if (rx>0.5 && rx<2.0)										//constraints fulfilled with acceptable stretching ratio
	{
		nx=count;
		rx=rx-((deltax_end-deltax_start)/fabs(deltax_end-deltax_start))*(mag_drx);	//to correct for the final increment/decrement in "rx"
		break;
	}											//solution is achieved only if rxmin1 = rxmin2
}
printf("\nThe requisite number of cells is %d and they need to be stretched by a factor of %f\n\n",nx,rx);
}


3.NSE-INPUT-EXTRACTION/FNPT_to_NS.CPP
//ROUTINE TO INTERPOLATE MOVING-GRID FNPT DATA ONTO A FIXED EULERIAN MESH FOR INPUT TO THE NAVIER-STOKES SOLVER 

#include <stdio.h>
#include <math.h>

main()
{
    FILE *f1;                   								//declaration of the file pointer for reading Output_PHIT.dat
    FILE *f2;                   								//declaration of the file pointer for mesh writing
    FILE *fU;                   								//declaration of the file pointer for writing to the Staggered U-mesh
    FILE *fW;                   								//declaration of the file pointer for writing to the Staggered W-mesh
    FILE *fp;                   								//declaration of the file pointer for writing to the Collocated p,f-mesh

    double d=1.0;               								//still water depth

    //input information for the NS domain
    double L=20.0;              								//length of the NS domain
    double H=2.0;               								//height of the NS domain
    int NX=500;                 								//total number of NS cells along x (has to be consistent with Global NS mesh size)
    int NZ=100;                 								//total number of NS cells along z (has to be consistent with Global NS mesh size)
    double dx=L/NX;             								//NS cell size along x
    double dz=H/NZ;             								//NS cell size along z

    //FNPT sub-domain extracted from IITM-FNPT2D - see "Output_PHIT.dat" and "FEMINPUT.dat"
    int NX_FNPT=10;             								//number of vertical FNPT layers along the x-direction [GRID IS FIXED]
    int NZ_FNPT=21;             								//number of horizontal FNPT layers along the z-direction [GRID MOVES VERTICALLY]

    //input information for the NS sub-domain where FNPT-data would be interpolated		/*...HARD-CODED - DO NOT CHANGE...*/
    int NX_NS=5;                								//number of cells along the length of the NS sub-domain
    int NZ_NS=NZ;               								//number of cells along the height of the NS sub-domain

    //number of time-steps executed in FNPT
    //N=(total no. of lines in "Output_PHIT.dat")/(NX_FNPT*NZ_FNPT)
    int N=5513;     										//verify against "FEMINPUT.dat" and "Output_PHIT.dat"

    //generation of the staggered U-mesh [back-staggered by 0.5*dx in the x-direction]
    double x_U[NX_NS+10]={0.0},z[NZ_NS+10]={0.0};
    x_U[1]=0.0;
    z[1]=0.5*dz;
    for (int i=2; i<=NX_NS; i++)								/***********************************************************************/
        x_U[i]=x_U[i-1]+dx;									/*...Change this routine if z-direction meshing is non-uniform in NS...*/
    for (int k=2; k<=NZ_NS; k++)								/***********************************************************************/
        z[k]=z[k-1]+dz;
    //writing the U-mesh to file (can be opened in TecPLOT)
    f2=fopen("Umesh.dat","w");
    fprintf (f2,"\nVARIABLES = ""x"", ""y""\n");
    fprintf (f2,"ZONE I=%d, J=%d, ZONETYPE=ORDERED, DATAPACKING=POINT\n",NZ_NS,NX_NS);
    for (int i=1; i<=NX_NS; i++)
        for (int k=1; k<=NZ_NS; k++)
            fprintf(f2,"%f\t%f\n",x_U[i],z[k]);
    fclose(f2);

    //generation of the collocated p,f-mesh
    double x[NX_NS+10]={0.0};
    x[1]=0.5*dx;
    z[1]=0.5*dz;
    for (int i=2; i<=NX_NS; i++)								/***********************************************************************/
        x[i]=x[i-1]+dx;										/*...Change this routine if z-direction meshing is non-uniform in NS...*/
    for (int k=2; k<=NZ_NS; k++)								/***********************************************************************/
        z[k]=z[k-1]+dz;
    //writing the p,f-mesh to file (can be opened in TecPLOT)
    f1=fopen("pmesh.dat","w");
    fprintf (f1,"\nVARIABLES = ""x"", ""y""\n");
    fprintf (f1,"ZONE I=%d, J=%d, ZONETYPE=ORDERED, DATAPACKING=POINT\n",NZ_NS,NX_NS);
    for (int i=1; i<=NX_NS; i++)
        for (int k=1; k<=NZ_NS; k++)
            fprintf(f1,"%f\t%f\n",x[i],z[k]);
    fclose(f1);

    //generation of the staggered W-mesh [back-staggered by 0.5*dz in the z-direction]
    double z_W[NZ_NS+10]={0.0};
    x[1]=0.5*dx;
    z_W[1]=0.0;
    for (int i=2; i<=NX_NS; i++)								/***********************************************************************/
        x[i]=x[i-1]+dx;										/*...Change this routine if z-direction meshing is non-uniform in NS...*/
    for (int k=2; k<=NZ_NS; k++)								/***********************************************************************/
        z_W[k]=z_W[k-1]+dz;
    //writing the W-mesh to file (can be opened in TecPLOT)
    f1=fopen("Wmesh.dat","w");
    fprintf (f1,"\nVARIABLES = ""x"", ""y""\n");
    fprintf (f1,"ZONE I=%d, J=%d, ZONETYPE=ORDERED, DATAPACKING=POINT\n",NZ_NS,NX_NS);
    for (int i=1; i<=NX_NS; i++)
        for (int k=1; k<=NZ_NS; k++)
            fprintf(f1,"%f\t%f\n",x[i],z_W[k]);
    fclose(f1);

    /*******************************************************************************************/	/**********************************/
    /*...Order of variables in Output_PHIT.dat : Time , X , Z , pressure , vx , vz , ax , az...*/	/*...HARD-CODED - DO NOT CHANGE...*/
    /*******************************************************************************************/	/**********************************/
    double FNPT_z[25]={0.0},FNPT_p[25]={0.0},FNPT_U[25]={0.0},FNPT_W[25]={0.0};
    f1=fopen("Output_ PHIT.dat","r");    							/*...open the FNPT output file in read mode...*/                    /*...KEEP IT OPEN...*/
    fU=fopen("Interp_Ustag.dat","w");   							/*...open a file to write U-velocity data over the U-grid...*/      /*...KEEP IT OPEN...*/
    fW=fopen("Interp_Wstag.dat","w");   							/*...open a file to write W-velocity data over the W-grid...*/      /*...KEEP IT OPEN...*/
    fp=fopen("Interp_pf.dat","w");      							/*...open a file to write p,f data over the collocated grid...*/    /*...KEEP IT OPEN...*/
    for (int count=1; count<=N; count++)    							//loop for all time steps [no time interpolation involved at this stage]
    {
        for (int i=1; i<=NX_FNPT/2; i++)    							//only half of the NX_FNPT nodes need to be considered for the U or pfW grids
        {
            /**************************************************************/
            /*...every (2*N-1) "FNPT_NZ lines" to be read to the U-grid...*/
            /**************************************************************/
            for (int k=1; k<=NZ_FNPT; k++)
            {
                fscanf(f1,"%*f\t%*f\t%lf\t%*f\t%lf\t%*f\t%*f\t%*f\n",&FNPT_z[k],&FNPT_U[k]); 	//only read z,p,vx,vz (skip the rest)
                FNPT_z[k]=FNPT_z[k]+d;  							//shifting the SWL to z=d
            }
                //*...interpolating FNPT information to the staggered U-grid...*//
                //*...interpolation is ONLY vertical...*//
                for (int l=1; l<=NZ_NS; l++)    						//vertical traverse over the staggered U-grid
                {
                    if (z[l]<FNPT_z[NZ_FNPT])   						//either a pure-cell in water or a mixed cell [U-cell center lies below the FNPT node]
                    {
                        for (int k=1; k<=NZ_FNPT; k++)
                        {
                        if (FNPT_z[k]>=z[l])
                        {
                            fprintf(fU,"%f\t%f\t%f\n",x_U[i],z[l],(((FNPT_U[k-1]*(FNPT_z[k]-z[l]))+(FNPT_U[k]*(z[l]-FNPT_z[k-1])))/(FNPT_z[k]-FNPT_z[k-1])));
                            break;
                        }
                        }
                    }
                    else                        						//either a pure-cell in air or a mixed cell [U-cell center lies above the FNPT node]
                            fprintf(fU,"%f\t%f\t%f\n",x_U[i],z[l],0.0);     			//surface velocity directly assigned
                }

            /*************************************************************************/
            /*...every (2*N) "FNPT_NZ lines" to be read to the W-grid and p,f-grid...*/
            /*************************************************************************/
            for (int k=1; k<=NZ_FNPT; k++)
            {
                fscanf(f1,"%*f\t%*f\t%lf\t%lf\t%*f\t%lf\t%*f\t%*f\n",&FNPT_z[k],&FNPT_p[k],&FNPT_W[k]); //only read z,p,vx,vz (skip rest)
                FNPT_z[k]=FNPT_z[k]+d;  							//shifting the SWL to z=d
            }
                //*...interpolating FNPT information to the staggered W-grid...*//
                //*...interpolation is ONLY vertical...*//
                for (int l=1; l<=NZ_NS; l++)    						//vertical traverse over the staggered W-grid
                {
                    if (z_W[l]<FNPT_z[NZ_FNPT]) 						//either a pure-cell in water or a mixed cell [W-cell center lies below the FNPT node]
                    {
                        for (int k=1; k<=NZ_FNPT; k++)
			{
                        if (FNPT_z[k]>=z_W[l])
                        {
			    if (FNPT_z[k]-FNPT_z[k-1]==0)
                            fprintf(fW,"%f\t%f\t%f\n",x[i],z_W[l],0.0);
			    else
                            fprintf(fW,"%f\t%f\t%f\n",x[i],z_W[l],(((FNPT_W[k-1]*(FNPT_z[k]-z_W[l]))+(FNPT_W[k]*(z_W[l]-FNPT_z[k-1])))/(FNPT_z[k]-FNPT_z[k-1])));
                            break;
                        }
			}
                    }
                    else                        						//either a pure-cell in air or a mixed cell [W-cell center lies above the FNPT node]
                            fprintf(fW,"%f\t%f\t%f\n",x[i],z_W[l],0.0);
                }
                //*...interpolating FNPT information to the collocated p,f-grid...*//
                //*...interpolation is ONLY vertical...*//
                for (int l=1; l<=NZ_NS; l++)    						//vertical traverse over the collocated p,f-grid
                {
                    if ((z[l]+0.5*dz)<=FNPT_z[NZ_FNPT])                                     	//pure cell lying entirely in water [p=rho*g*z and f=1]
                    {
                        for (int k=1; k<=NZ_FNPT; k++)
			{
                        if (z[l]>=FNPT_z[k])
                        {
                            fprintf(fp,"%f\t%f\t%f\t%f\n",x[i],z[l],1000.0*(((FNPT_p[k]*(FNPT_z[k+1]-z[l]))+(FNPT_p[k+1]*(z[l]-FNPT_z[k])))/(FNPT_z[k+1]-FNPT_z[k])),1.0);
                            break;
                        }
			}
                    }
                    else if (fabs(z[l]-FNPT_z[NZ_FNPT])<0.5*dz && FNPT_z[NZ_FNPT]>z[l])     	//mixed cell with the FNPT node lying above the p,f-cell center
                    {
                            int k=NZ_FNPT-1;
                            fprintf(fp,"%f\t%f\t%f\t%f\n",x[i],z[l],1000.0*(((FNPT_p[k]*(FNPT_z[k+1]-z[l]))+(FNPT_p[k+1]*(z[l]-FNPT_z[k])))/(FNPT_z[k+1]-FNPT_z[k])),(((FNPT_z[NZ_FNPT]-z[l])+0.5*dz)/dz));
                    }
                    else if (fabs(z[l]-FNPT_z[NZ_FNPT])<0.5*dz && FNPT_z[NZ_FNPT]<z[l])     	//mixed cell with the FNPT node lying below the p,f-cell center
                            fprintf(fp,"%f\t%f\t%f\t%f\n",x[i],z[l],0.0,(((FNPT_z[NZ_FNPT]-z[l])+0.5*dz)/dz));
                    else                                                                    	//pure cell lying entirely in air [p=0.0 and f=0.0]
                            fprintf(fp,"%f\t%f\t%f\t%f\n",x[i],z[l],0.0,0.0);
                }
        }//end i-loop
    }//end time-loop
    fclose(f1);
    fclose(fU);
    fclose(fW);
    fclose(fp);
}
